pkg/util/kubernetes/kubelet/kubelet.go

```diff
@@ -497,44 +499,289 @@ func (ku *KubeUtil) setupKubeletApiEndpoint() error {
+		}
+	}
+
+	return ips, hostnames
+}
+
+func potentialKubeletHostsFilter(h *host) {
+	ipsKeys := make(map[string]bool)
+	ips := []string{}
+	for _, ip := range h.ips {
+		if _, check := ipsKeys[ip]; !check {
+			ipsKeys[ip] = true
+			ips = append(ips, ip)
+		}
+	}
+
+	hostnamesKeys := make(map[string]bool)
+	hostnames := []string{}
+	for _, hostname := range h.hostnames {
+		if _, check := hostnamesKeys[hostname]; !check {
+			hostnamesKeys[hostname] = true
+			hostnames = append(hostnames, hostname)
+		}
+	}
+
+	h.ips = ips
+	h.hostnames = hostnames
+}
+
+func (ku *KubeUtil) setKubeletHost(h *host, httpsPort, httpPort int) error {
+	var err error
+	for _, ip := range h.ips {
+		log.Infof("Trying to use host %s with HTTPS", ip)
+		ku.kubeletHost = ip
+		err := ku.setupKubeletApiClient()
+		if err != nil {
+			return err
+		}
+
+		err = checkKubeletHTTPSConnection(ku, httpsPort)
+		if err == nil {
+			log.Infof("Can connect to kubelet using %s and HTTPS, setting up %s as kubelet host", ip, ip)
+			ku.kubeletHost = ip
+			return nil
+		}
+	}
+
+	for _, hostname := range h.hostnames {
+		log.Infof("Trying to use host %s with HTTPS", hostname)
+		ku.kubeletHost = hostname
+		err := ku.setupKubeletApiClient()
+		if err != nil {
+			return err
+		}
+
+		err = checkKubeletHTTPSConnection(ku, httpsPort)
+		if err == nil {
+			log.Infof("Can connect to kubelet using %s and HTTPS, setting up %s as kubelet host", hostname, hostname)
+			ku.kubeletHost = hostname
+			return nil
+		}
+	}
+
+	for _, ip := range h.ips {
+		log.Infof("Trying to use host %s with HTTP", ip)
+		err = checkKubeletHTTPConnection(ip, httpPort)
+		if err == nil {
+			log.Infof("Can connect to kubelet using %s and HTTP, setting up %s as kubelet host", ip, ip)
+			ku.kubeletHost = ip
+			return nil
+		}
+	}
+
+	for _, hostname := range h.hostnames {
+		log.Infof("Trying to use host %s with HTTP", hostname)
+		err = checkKubeletHTTPConnection(hostname, httpPort)
+		if err == nil {
+			log.Infof("Can connect to kubelet using %s and HTTP, setting up %s as kubelet host", hostname, hostname)
+			ku.kubeletHost = hostname
+			return nil
+		}
+	}
+
+	log.Warnf("Cannot set a valid kubeletHost: cannot connect to kubelet using any of the given hosts: %v %v", h.ips, h.hostnames)
+	ku.kubeletHost = config.Datadog.GetString("kubernetes_kubelet_host")
+	return nil
+}
+
+func checkKubeletHTTPSConnection(ku *KubeUtil, httpsPort int) error {
 	c := http.Client{Timeout: time.Second}
 	c.Transport = &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}
+	log.Debugf("Trying to query the kubelet endpoint %s ...", ku.kubeletApiEndpoint)
+	ku.kubeletApiEndpoint = fmt.Sprintf("https://%s:%d", ku.kubeletHost, httpsPort)
+	response, err := c.Get(ku.kubeletApiEndpoint + "/")
+	if err == nil {
+		log.Infof("Successfully queried to %s without any security settings, adding security transport settings to query %s%s", response.Request.URL, ku.kubeletApiEndpoint, kubeletPodPath)
```

Comment:
nitpick: "Successfully queried ~to~ %s"

Reply:
sorry for the spelling mistakes, i'll be more careful from now on